File: arrays_hashing/twoSum.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        numMap = {}
        for i,val in enumerate(nums):
            remainder = target - val
            if remainder in numMap:
                indexOfRemainder = numMap[remainder]
                return [indexOfRemainder, i]

            if val not in numMap:
                numMap[val] = i
                

def main():
    s = Solution()
    print(s.twoSum([2,7,11,15], 9))
    print(s.twoSum([3,2,4], 6))
    print(s.twoSum([3,3], 6))

if __name__ == "__main__":
    main()

File: arrays_hashing/twoSum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int,int> numMap;
        vector<int> ret;

        for (int i = 0; i<nums.size(); i++){
            if(numMap.count(target - nums[i])){
                ret = {numMap[target - nums[i]], i};
                return ret;
            }

            if (!numMap.count(nums[i])){
                numMap[nums[i]] = i;
            }

        }
        return ret;
    }
};

int main(){
    Solution s;
    vector<int> nums = {2,7,11,15};
    vector<int> ret = s.twoSum(nums,9);
    for (auto& val : ret){
        cout << val << " ";
    }
    cout << endl;
    return 0;
}